compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -1265,24 +1266,89 @@ public void visit(BLangAssignment assignNode) {
     }
 
     public void visit(BLangRecordDestructure stmt) {
+        List<BLangExpression> varRefs = stmt.varRef.recordRefFields.stream()
+                .map(e -> e.variableReference).collect(Collectors.toList());
+        varRefs.add((BLangExpression) stmt.varRef.restParam);
+        this.checkDuplicateVarRefs(varRefs);
         this.checkStatementExecutionValidity(stmt);
         analyzeExpr(stmt.varRef);
         analyzeExpr(stmt.expr);
     }
 
     public void visit(BLangErrorDestructure stmt) {
+        List<BLangExpression> varRefs = new ArrayList<>();
+        varRefs.add(stmt.varRef.reason);
+        varRefs.addAll(stmt.varRef.detail.stream().map(e -> e.expr).collect(Collectors.toList()));
+        varRefs.add(stmt.varRef.restVar);
+        this.checkDuplicateVarRefs(varRefs);
         this.checkStatementExecutionValidity(stmt);
         analyzeExpr(stmt.varRef);
         analyzeExpr(stmt.expr);
     }
 
     @Override
     public void visit(BLangTupleDestructure stmt) {
+        List<BLangExpression> varRefs = new ArrayList<>(stmt.varRef.expressions);
+        varRefs.add((BLangExpression) stmt.varRef.restParam);
+        this.checkDuplicateVarRefs(varRefs);
         this.checkStatementExecutionValidity(stmt);
         analyzeExpr(stmt.varRef);
         analyzeExpr(stmt.expr);
     }
 
+    private void checkDuplicateVarRefs(List<BLangExpression> varRefs) {
+        checkDuplicateVarRefs(varRefs, new HashSet<>());
+    }
+
+    private void checkDuplicateVarRefs(List<BLangExpression> varRefs, Set<BSymbol> symbols) {
+        for (BLangExpression varRef : varRefs) {
+            if (varRef == null || (varRef.getKind() != NodeKind.SIMPLE_VARIABLE_REF
+                    && varRef.getKind() != NodeKind.INDEX_BASED_ACCESS_EXPR
+                    && varRef.getKind() != NodeKind.FIELD_BASED_ACCESS_EXPR
+                    && varRef.getKind() != NodeKind.XML_ATTRIBUTE_ACCESS_EXPR
```

Comment:
NodeKind.INDEX_BASED_ACCESS_EXPR and NodeKind.FIELD_BASED_ACCESS_EXPR and XML_ATTRIBUTE_ACCESS_EX are not allowed according to the spec.  Let's fix them in a separate PR.

Reply:
Created https://github.com/ballerina-platform/ballerina-lang/issues/18249